src/fastpath/fastpath.c

```diff
@@ -6,15 +6,275 @@
+        slowpath(SysSend);
+    }
+
+    /* Get vspace root. */
+    cap_pd = cap_vtable_cap_get_vspace_root_fp(newVTable);
+
+#ifdef CONFIG_ARCH_AARCH32
+    /* Get HW ASID */
+    stored_hw_asid = cap_pd[PD_ASID_SLOT];
+#endif
+
+#ifdef CONFIG_ARCH_X86_64
+    /* borrow the stored_hw_asid for PCID */
+    stored_hw_asid.words[0] = cap_pml4_cap_get_capPML4MappedASID_fp(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_IA32
+    /* stored_hw_asid is unused on ia32 fastpath, but gets passed into a function below. */
+    stored_hw_asid.words[0] = 0;
+#endif
+
+#ifdef CONFIG_ARCH_AARCH64
+    stored_hw_asid.words[0] = cap_vtable_root_get_mappedASID(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_RISCV
+    /* Get HW ASID */
+    stored_hw_asid.words[0] = cap_page_table_cap_get_capPTMappedASID(newVTable);
+#endif
+
+    // TODO: Needed?
+    /* Try and wake up threads - this makes everything fail */
+    // awaken();
+
+    /* Let gcc optimise this out for 1 domain */
+    dom = maxDom ? ksCurDomain : 0;
+    /* Ensure only the idle thread or lower prio threads are present in the scheduler */
+    if (unlikely(dest->tcbPriority < NODE_STATE(ksCurThread->tcbPriority) &&
+                 !isHighestPrio(dom, dest->tcbPriority))) {
+
+        slowpath(SysSend);
+    }
+
+#ifdef CONFIG_ARCH_AARCH32
+    if (unlikely(!pde_pde_invalid_get_stored_asid_valid(stored_hw_asid))) {
+        slowpath(SysSend);
+    }
+#endif
+    /* Ensure the destination thread is in the current domain and can be scheduled directly. */
+    if (unlikely(dest->tcbDomain != ksCurDomain && maxDom)) {
+        slowpath(SysSend);
+    }
+#ifdef ENABLE_SMP_SUPPORT
+    /* Ensure both threads have the same affinity */
+    if (unlikely(NODE_STATE(ksCurThread)->tcbAffinity != dest->tcbAffinity)) {
+        slowpath(SysSend);
+    }
+#endif /* ENABLE_SMP_SUPPORT */
+    // TODO: Needed?
+    /* This is basically the check in checkDomainTime(), which needs more scheduler logic */
+    if (unlikely(isCurDomainExpired())) {
+        slowpath(SysSend);
+    }
+
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+    ksKernelEntry.is_fastpath = true;
+#endif
+
+    thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+    updateTimestamp();
+    ticks_t prev = getNextInterrupt();
+
+    if (checkBudget()) {
+        commitTime();
+    }
+
+    if (isSchedulable(NODE_STATE(ksCurThread))) {
+        SCHED_ENQUEUE_CURRENT_TCB;
+    }
+
+    switchToThread_fp(dest, cap_pd, stored_hw_asid);
+
+    ticks_t next = getNextInterrupt();
+    if (next < prev) {
+        setDeadline(next - getTimerPrecision());
+    }
+
+    // TODO: Something like this is done in the slowpath - unsure if needed in fastpath.
+    if (sc_sporadic(dest->tcbSchedContext)) {
+        refill_unblock_check(dest->tcbSchedContext);
+    }
+
+    NODE_STATE(ksCurSC) = NODE_STATE(ksCurThread)->tcbSchedContext;
+
+    /* TODO: In case NODE_STATE(ksSchedulerAction) was messed up by awaken(). Is awaken needed though? */
```

Comment:
Same as above: if the fastpath allows conditions under which `awaken` is called on the slowpath is has to be in. Instead of pulling in more, though, the ides is usually to tighten fastpath constraints, otherwise you end up just running the slowpath.

Reply:
The reason this is a little tricky is because even in the seL4_Call/ seL4_ReplyRecv fastpaths, this _should_ be run - as it is a non-optional part of `schedule()`, which is run in `handleSyscall()` for the slowpath. I assume this is omitted in the fastpath because there is no switching of scheduling contexts, but this is not the case for this version of the signal fastpath. I'm not aware of any fastpath where a signalling context change happens, which is why I had some difficulty narrowing down exactly what could and couldn't be omitted.

Reply:
@alwin-joshy this is why I did not fastpath the case that needs awaken.